sdk/rust/oak/src/handle.rs

```diff
@@ -0,0 +1,224 @@
+/// [`oak::io::Sender`](../io/struct.Sender.html).
+///
+/// This trait can be automatically derived:
+/// ```
+/// use oak::handle::HandleVisit;
+///
+/// #[derive(HandleVisit)]
+/// struct Thing {
+///     receiver: oak::io::Receiver<String>,
+///     sender: oak::io::Sender<String>,
+/// }
+/// ```
+///
+/// Alternatively, you can implement the trait manually. This is required if you want the trait to
+/// visit handles that are directly contained in a type.
+///
+/// ```
+/// use oak::handle::{Handle, HandleVisit};
+///
+/// struct Thing {
+///     handle: Handle,
+/// }
+///
+/// impl HandleVisit for Thing {
+///     fn visit<F: FnMut(&mut Handle)>(&mut self, mut visitor: F) -> F {
+///         visitor(&mut self.handle);
+///         visitor
+///     }
+/// }
+/// ```
+pub trait HandleVisit {
+    /// Invokes the provided closure on every handle contained in `self`.
+    fn visit<F: FnMut(&mut Handle)>(&mut self, visitor: F) -> F;
+}
+
+// A default implementation of the HandleVisit trait that does nothing
+macro_rules! handle_visit_blanket_impl {
+    ($($t:ty),+) => {
+        $(
+            impl HandleVisit for $t {
+                fn visit<F: FnMut(&mut Handle)>(&mut self, visitor: F) -> F {
+                    visitor
+                }
+            }
+        )+
+    };
+}
+
+// Provide an implementation for all scalar types in Prost.
+// See: https://github.com/danburkert/prost#scalar-values
+handle_visit_blanket_impl!((), f64, f32, i32, i64, u32, u64, bool, String, Vec<u8>);
+
+// Provide an implementation for oak_abi types that an implementation cannot be derived for.
+// These do not contains handles, so a blanket impl is sufficient.
+handle_visit_blanket_impl!(
+    oak_abi::proto::oak::encap::GrpcResponse,
+    oak_abi::proto::oak::encap::GrpcRequest,
+    oak_abi::proto::oak::log::LogMessage,
+    oak_abi::proto::oak::application::ConfigMap
+);
+
+/// Return all handles in `T`.
+///
+/// Also marks all handles in `T` invalid.
+///
+/// The original message can be reconstructed by calling
+/// [`inject_handles`](fn.inject_handles.html).
+///
+/// ```
+/// use oak::handle::{extract_handles, Handle, HandleVisit};
+///
+/// struct Thing {
+///     handle: Handle,
+/// }
+/// # impl HandleVisit for Thing {
+/// #   fn visit<F: FnMut(&mut Handle)>(&mut self, mut visitor: F) -> F {
+/// #     visitor(&mut self.handle);
+/// #     visitor
+/// #   }
+/// # }
+///
+/// let mut thing = Thing { handle: 42 };
+///
+/// let handles = extract_handles(&mut thing);
+///
+/// assert_eq!(handles, vec![42]);
+/// ```
+pub fn extract_handles<T: HandleVisit>(msg: &mut T) -> Vec<Handle> {
+    let mut handles = Vec::new();
+    msg.visit(|handle: &mut Handle| {
+        handles.push(*handle);
+        *handle = oak_abi::INVALID_HANDLE;
+    });
+    handles
+}
+
+/// Inject handles into a message.
+///
+/// If the number of handles provided is not exactly equal to the number of handles needed to fill
```

Comment:
Could maybe do with some more text here pointing out that the order of the `handles` also needs to be the same as the order of visitation, or else things will (silently) go awry.

"I'm playing all the right notes, but not necessarily in the right order"